packages/realm-react/src/useObject.tsx

```diff
@@ -55,24 +55,94 @@ export function createUseObject(useRealm: () => Realm) {
+    const primaryKeyRef = useRef(primaryKey);
+
+    // Initializing references with a function call or class constructor will
+    // cause the function or constructor to be called on ever render.
+    // Even though this value is thrown away, `createCachedObject` will end up registering a listener.
+    // Therefore, we initialize the references with null, and only create the object if it is null
+    // Ref: https://github.com/facebook/react/issues/14490
+    const cachedObjectRef = useRef<null | CachedObject>(null);
+
+    if (cachedObjectRef.current === null) {
+      cachedObjectRef.current = createCachedObject({
+        object: originalObject ?? null,
+        realm,
+        updateCallback: forceRerender,
+      });
+    }
+
+    // Create a ref, since the object returned from `objectForPrimaryKey` is always going to have a different reference
+    const originalObjectRef = useRef(originalObject);
+
     // Wrap the cachedObject in useMemo, so we only replace it with a new instance if `primaryKey` or `type` change
     const { object, tearDown } = useMemo(
       // TODO: There will be an upcoming breaking change that makes objectForPrimaryKey return null
       // When this is implemented, remove `?? null`
-      () =>
-        createCachedObject({
-          object: realm.objectForPrimaryKey<unknown>(type, primaryKey) ?? null,
-          realm,
-          updateCallback: forceRerender,
-        }),
-      [type, realm, primaryKey],
+      () => {
+        // This should never happen, but if it does, we want to return a null result
+        if (!cachedObjectRef.current) {
+          return { object: null, tearDown: () => undefined };
+        }
+
+        // Re-instantiate the cachedObject if the primaryKey has changed or the originalObject has gone from null to not null
+        if (
+          comparePrimaryKeys(primaryKey, primaryKeyRef.current) === false ||
+          (originalObjectRef.current === null && originalObject !== null)
+        ) {
+          cachedObjectRef.current = createCachedObject({
+            object: originalObject ?? null,
+            realm,
+            updateCallback: forceRerender,
+          });
+          originalObjectRef.current = originalObject;
+        }
+        return cachedObjectRef.current;
+      },
+      [realm, originalObject, primaryKey],
     );
 
     // Invoke the tearDown of the cachedObject when useObject is unmounted
     useEffect(() => {
       return tearDown;
     }, [tearDown]);
 
+    // If the object doesn't exist, listen for insertions to the collection and force a rerender if the inserted object has the correct primary key
+    useEffect(() => {
+      const collectionListener: Realm.CollectionChangeCallback<T & Realm.Object> = (collection, changes) => {
+        for (const index of changes.insertions) {
+          const object = collection[index];
+          const primaryKeyProperty = object.objectSchema().primaryKey;
```

Comment:
This seem costly - is there any way this could be reused across items? They all share the same object schema.